### PR TITLE
Improve IBBrokerage Symbol Mapping Error Message

### LINKED ISSUE #5012: Improve Exception Message for CreateHolding in InteractiveBrokersBrokerage
#### Expected Behavior
In the exception message, the information about the symbol that the user holds in his IBKR account should be present. E.g.:
"Failed to convert holding index-options RUT, this is not supported by LEAN"

#### Actual Behavior
The algorithm doesn't give any information about the failed conversion. If gives, for example, 
"String was not recognized as a valid DateTime." at `InteractiveBrokersBrokerage.MapSymbol` method.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

Use a try catch around IBBrokerages symbol mapping to better clarify what contract is breaking the conversion.

#### Related Issue

Closes #5012 

#### Motivation and Context

Reference issue #5012 

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

Failed to replicate the issue in testing but at the very least the new try catch can give us some insight in the future as to what the problem may be stemming from.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 1]
```diff
@@ -2244,54 +2244,61 @@ private Holding CreateHolding(IB.UpdatePortfolioEventArgs e)
...
-            if (securityType == SecurityType.Option)
+            catch (Exception error)
             {
-                Log.Trace($"Mapping contract {contract} with ticker `{ibSymbol}` and SecurityType `{securityType.ToString()}` with expiry: {contract.LastTradeDateOrContractMonth} to Symbol");
-                var expiryDate = DateTime.ParseExact(contract.LastTradeDateOrContractMonth, DateFormat.EightCharacter, CultureInfo.InvariantCulture);
-                var right = contract.Right == IB.RightType.Call ? OptionRight.Call : OptionRight.Put;
-                var strike = Convert.ToDecimal(contract.Strike);
-
-                return _symbolMapper.GetLeanSymbol(ibSymbol, securityType, market, expiryDate, strike, right);
+                throw new Exception($"InteractiveBrokersBrokerage.MapSymbol(): Failed to convert contract for {contract.Symbol}; Contract description: {GetContractDescription(contract)}; Error: {error.Message}");
```
Comment (gsalaz98) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs:2300]: Let's log and rethrow the original exception instead. We lose the context of the original exception by rethrowing with a new exception.

Try:

```C#
Log.Error(msg);
throw;
```

instead 👍

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 2]
```diff
@@ -2244,54 +2244,61 @@ private Holding CreateHolding(IB.UpdatePortfolioEventArgs e)
...
 
-                // Create a canonical future Symbol for lookup in the FuturesExpiryFunctions helper class.
-                // We then get the delta between the futures option's expiry month vs. the future's expiry month.
-                var canonicalFutureSymbol = _symbolMapper.GetLeanSymbol(ibSymbol, SecurityType.Future, market, SecurityIdentifier.DefaultDate);
-                var futureExpiryFunction = FuturesExpiryFunctions.FuturesExpiryFunction(canonicalFutureSymbol);
-                var futureContractExpiryDate = futureExpiryFunction(FuturesOptionsExpiryFunctions.GetFutureContractMonth(canonicalFutureSymbol, contractExpiryDate));
-                var futureSymbol = Symbol.CreateFuture(canonicalFutureSymbol.ID.Symbol, canonicalFutureSymbol.ID.Market, futureContractExpiryDate);
+                if (securityType == SecurityType.Option)
+                {
+                    Log.Trace($"Mapping contract {contract} with ticker `{ibSymbol}` and SecurityType `{securityType.ToString()}` with expiry: {contract.LastTradeDateOrContractMonth} to Symbol");
```
Comment (gsalaz98) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs:2288]: I think we can remove this statement now that the exception will print out this data for us

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 3]
```diff
@@ -2244,54 +2244,61 @@ private Holding CreateHolding(IB.UpdatePortfolioEventArgs e)
...
+                    var leanSymbol = _symbolMapper.GetLeanRootSymbol(ibSymbol);
+                    var defaultMarket = market;
+                    if (!_symbolPropertiesDatabase.TryGetMarket(leanSymbol, SecurityType.Future, out market))
+                    {
+                        market = defaultMarket;
+                    }
 
-                Log.Trace($"Mapping contract {contract} with ticker `{leanSymbol}` and SecurityType `{securityType.ToString()}` with expiry: {contract.LastTradeDateOrContractMonth} to Symbol");
-                var contractExpiryDate = DateTime.ParseExact(contract.LastTradeDateOrContractMonth, DateFormat.EightCharacter, CultureInfo.InvariantCulture);
+                    Log.Trace($"Mapping contract {contract} with ticker `{leanSymbol}` and SecurityType `{securityType.ToString()}` with expiry: {contract.LastTradeDateOrContractMonth} to Symbol");
```
Comment (gsalaz98) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs:2265]: We can remove this now 👍

**Review by gsalaz98 — CHANGES_REQUESTED** [review]
The data triggering this exception is because we're trying to parse a DateTime from the value `0`. The TWS API docs attached say that data may be returned as `0` whenever too much data is requested. I wonder if the two are related... 

image

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**